server/controllers/images.ts

```diff
@@ -0,0 +1,145 @@
+import path from 'path';
+
+import { ManagedUpload } from 'aws-sdk/clients/s3';
+import { encode } from 'blurhash';
+import config from 'config';
+import { kebabCase } from 'lodash';
+import sharp from 'sharp';
+import { v1 as uuid } from 'uuid';
+
+import s3, { uploadToS3 } from '../lib/awsS3';
+import errors from '../lib/errors';
+import { isSupportedImageMimeType } from '../lib/images';
+import { reportErrorToSentry } from '../lib/sentry';
+import models from '../models';
+import { SUPPORTED_FILE_EXTENSIONS, SUPPORTED_FILE_KINDS, SUPPORTED_FILE_TYPES } from '../models/UploadedFile';
+
+// Use a 2 minutes timeout for image upload requests as the default 25 seconds
+// often leads to failing requests.
+const IMAGE_UPLOAD_TIMEOUT = 2 * 60 * 1000;
+
+const getUploadedFileData = async file => {
+  if (isSupportedImageMimeType(file.mimetype)) {
+    const image = sharp(file.buffer);
+    const { width, height } = await image.metadata();
+    const pixels = await image.raw().ensureAlpha().toBuffer({ resolveWithObject: true });
+    const blurHash = encode(pixels.data, width, height, 4, 4); // TODO find optimal values
+    return { width, height, blurHash };
+  } else {
+    return null;
+  }
+};
+
+const recordUploadedFile = async (
+  req,
+  kind,
+  file,
+  fileName,
+  s3Data: ManagedUpload.SendData | { Location: string; Bucket: string; Key: string },
+): Promise<void> => {
+  try {
+    await models.UploadedFile.create({
+      kind: kind,
+      fileName,
+      fileSize: file.size,
+      fileType: file.mimetype,
+      url: s3Data.Location,
+      data: await getUploadedFileData(file),
+      CreatedByUserId: req.remoteUser.id,
+    });
+  } catch (e) {
+    // Since this runs async, we need to report the error to Sentry in case it fails
+    console.error(e);
+    reportErrorToSentry(e, {
+      severity: 'error',
+      user: req.remoteUser,
+      transactionName: 'recordUploadedFile',
+      extra: { kind, fileName, fileType: file.mimetype },
+    });
+  }
+};
+
+const getFilename = (file, fileNameFromArgs) => {
+  const expectedExtension = SUPPORTED_FILE_EXTENSIONS[file.mimetype];
+  const rawFileName = fileNameFromArgs || file.originalname || uuid();
```

Comment:
Do you think we should limit the filename size input? s3 object keys are limited to 1024 bytes.

Reply:
Let's do that